### ISSUE TITLE
Get rid of OPL emulations other than NukedOPL

### ISSUE THREAD

**Issue by johnnovak** [dosbox-staging/dosbox-staging#1723]
What the title says 😄 NukedOPL offers the best emulation accuracy, it provides bit-perfect output compared to the real chips. There's little sense in carrying the other much inferior OPL emulations forward (fast, compat, and mame); they're just a maintenance burden and I can't imagine anyone preferring them over NukedOPL. The difference in performance is a non-issue on modern hardware.

**Comment by Burrito78** [comment]
Wasn't it that NukedOPL can be a little bit too resource hungry for Raspberry Pi systems? @kcgen knows that, he did talk about this at some point.

**Comment by kcgen** [comment]
> Wasn't it that NukedOPL can be a little bit too resource hungry for Raspberry Pi systems? @kcgen knows that, he did talk about this at some point.

Yes; when I captured a perf snapshot on the Pi3, Nuked uses almost exactly twice the amount of host CPU versus `oplemu = fast` (the later does indeed live up to its name!)

But the good news is that it's only a couple percent more when taking into account total DOSBox CPU needs, even at low cycle counts like 3k.

**Comment by kcgen** [comment]
> Also we have to check with Team eXoDOS if they need any of the old emulators for some obscure reason.

Ping sent.

**Comment by johnnovak** [comment]
Yeah, and on any desktop CPU released in the past 10 years the difference is even more negligible.

Based on my understanding this is the relative ranking of the various OPL emulations in terms of accuracy:

```
mame <<<<< fast << compat <<<<<<<<<<< nuked
```

Basically nuked beats everything else, it's bit-perfect compared to the original hardware. I read it somewhere that at some point the guy started looking at decapped images of the sound chip, so it should be pretty accurate even when dealing with some wacky unspecified input combinations.

Therefore, I can't imagine any use case for the earlier OPL emulation modes if the performance differences don't matter anymore (remember, compat is from around 2002 and fast was introduced in the early days of DOSBox, the performance difference might have mattered more then).

**Comment by kcgen** [comment]
> > Also we have to check with Team eXoDOS if they need any of the old emulators for some obscure reason.

Thanks to @Python-Exoproject for confirming:
 - We override `oplemu = ` and use Nuked for games using opl3, DualOPL2, or opl3gold, otherwise we default
 - We have one game that uses `oplemu = fast`, however there are no notes on why its set that way. I've confirmed its not needed. So no, we have no games that need anything special

**Comment by sulix** [comment]
Just to quickly jump in with one related thought — having some way to select the OPL implementation will probably be required to support real OPL2 chips, either on soundcards which still have them, or using the OPL{2,3}LPT or retrowave boards. So even if other emulators are removed, it may be worth keeping _some_ amount of indirection/configurability here.

(I'm working on [a library](https://github.com/sulix/liboplhw) to support these, and have an old [draft patch for DOSBox](https://davidgow.net/files/0001-dosbox-liboplhw.patch) to use it.)

**Comment by johnnovak** [comment]
I've done some digging and it turns out ScummVM has added support not only for Retrowave OPL, but also something called OPL2LPT this year (OPL-over-parallel-port).

The good news is that both implementations are very simple, short, and non-invasive:

https://github.com/scummvm/scummvm/blob/af85c911a5d660da911a45aae62b77ade2376860/audio/rwopl3.cpp
https://github.com/scummvm/scummvm/blob/af85c911a5d660da911a45aae62b77ade2376860/audio/opl2lpt.cpp

They both completely bypass SDL for the I/O, but do so in a clean cross-platform fashion. The Retrowave OPL code is from DOSBox-X; it uses the cross-platform Retrowave SDK for I/O which only needs some OS-specific `ifdefs` around the init, but that's no big deal. The SDK is supposed to be fully compatible with the Raspberry Pi.

The OPL2LPT stuff uses `libieee1284` which is again cross-platform.

I think @kcgen and myself are on the same page about this: while interfacing with niche hardware is definitely outside the scope and mission goal of the project, adding this would certainly be a cool little feature. I think I'm personally fine with it *as long as* it can be done cleanly for all supported platforms, and it wouldn't impose an unreasonable maintenance burden — which seems like the case. And we don't want to lag behind ScummVM either 😏 

I was also considering that only people who own such niche hardware would be able to maintain this code path, but given its self-contained and simple nature, I don't think that will be a problem. Seems like a case of write-once-and-forget code (makes sense as the hardware itself is fixed).

I'll still strip down the current OPL implementation to its simplest form, then you'll need to add some minimal abstraction layer on top of that when you'll get to this. There's a few things to consider that I'll leave to you @sulix (e.g. rendering wouldn't happen with a hardware device). To be honest, I think handling the various config options and initialising things correctly will be more complicated than introducing that abstraction (e.g. the OPL init stuff is tightly coupled with the Sound Blaster code, then there's the `auto` selection of a suitable OPL model for the current `sbtype`, etc.)

**Comment by sulix** [comment]
That sounds like a plan: As an SDL developer, I can definitely sympathise with the goal of relying on it for all the platform interfacing possible.

My personal plan is to have all of the real-OPL-hardware support in [a library](, which can then be shared between any program which would benefit, from dosbox-staging to game source-ports. I've got support for the OPL2LPT, Retrowave OPL, and the Linux kernel's OPL2 driver stack (for, e.g., real OPL2 soundcards), but it's all Linux-specific, and OPL2-only for now, so other OSes and OPL3 support are still a work-in-progress. The only config option it needs thus far, though, is a "device path", such as "retrowave:/dev/ttyACM0" or "alsa:", which can sometimes be auto-detected or read from an environment variable.

Regardless, I'll have a look at how best to add something like this once the stripped down NukedOPL implementation is done — as you say, the most difficult bit is likely to be the config handling.

**Comment by kcgen** [comment]
Adding a private OPL member function: `void PrintDiagnostics(const io_port_t base) const;`:

``` c++
void OPL::PrintDiagnostics(const io_port_t base) const
{
	// lamda to convert OplMode to char* string
	auto mode_to_string = [](const Mode m) {
		switch (m) {
		case Mode::OPL2: return "OPL2";
		case Mode::DualOPL2: return "DualOPL2";
		case Mode::OPL3: return "OPL3";
		case Mode::OPL3Gold: return "OPL3Gold";
		}
		return "Unknown";
	};
	LOG_INFO("OPL: Diagnostics");
	LOG_INFO(" |- Mode: %s", mode_to_string(mode));

	// lamda to read a port and return a string if the response is valid
	auto get_port_state = [](const io_port_t p) {
		return IO_ReadB(p) != 0xff ? "answering" : "not answering";
	};

	LOG_INFO(" |- Port 388h: %s", get_port_state(0x388));

	if (base) {
		LOG_INFO(" |- Port %3xh is: %s", base, get_port_state(base));
		LOG_INFO(" |- Port %3xh is: %s", base + 8, get_port_state(base + 8));
	}

	static auto prev_base = io_port_t{0};
	if (prev_base) {
		LOG_INFO(" |- (prior) Port %3xh is: %s",
		         prev_base,
		         get_port_state(prev_base));
		LOG_INFO(" |- (prior) Port %3xh is: %s",
		         prev_base + 8,
		         get_port_state(prev_base + 8));
	}
	prev_base = base;

	LOG_INFO(" |- DualOPL2 bit (newm): %u", newm);
	LOG_INFO(" |- AdlibGold smart pointer: %s",
	         adlib_gold ? "populated" : "nullptr");
	LOG_INFO(" `- channel smart pointer: %s",
	         mixer_chan ? "populated" : "nullptr");
	if (mixer_chan) {
		LOG_INFO("      |- is enabled: %s",
		         mixer_chan->is_enabled ? "yes" : "no");
		LOG_INFO("      `- sample rate: %d Hz", mixer_chan->GetSampleRate());
	}
}
```

Called at the end of the constructor:

``` c++
OPL::OPL(Section *configuration)
{
	/// ...

	PrintDiagnostics(base);
}
```


---

Looks like all is well on state changes:

#### `Z:\> oplmode opl2`

``` 
2022-06-21 06:44:07.549 | OPL: Diagnostics
2022-06-21 06:44:07.549 |  |- Mode: OPL2
2022-06-21 06:44:07.549 |  |- Port 388h: answering
2022-06-21 06:44:07.549 |  |- Port 220h is: not answering
2022-06-21 06:44:07.549 |  |- Port 228h is: answering
2022-06-21 06:44:07.549 |  |- (prior) Port 220h is: not answering
2022-06-21 06:44:07.549 |  |- (prior) Port 228h is: answering
2022-06-21 06:44:07.549 |  |- DualOPL2 bit (newm): 0
2022-06-21 06:44:07.549 |  |- AdlibGold smart pointer: nullptr
2022-06-21 06:44:07.549 |  `- channel smart pointer: populated
2022-06-21 06:44:07.549 |       |- is enabled: no
2022-06-21 06:44:07.549 |       `- sample rate: 48000 Hz
```

#### `Z:\> oplmode dualopl2`

```
2022-06-21 06:44:14.141 | OPL: Diagnostics
2022-06-21 06:44:14.141 |  |- Mode: DualOPL2
2022-06-21 06:44:14.141 |  |- Port 388h: answering
2022-06-21 06:44:14.141 |  |- Port 220h is: answering
2022-06-21 06:44:14.141 |  |- Port 228h is: answering
2022-06-21 06:44:14.141 |  |- (prior) Port 220h is: answering
2022-06-21 06:44:14.142 |  |- (prior) Port 228h is: answering
2022-06-21 06:44:14.142 |  |- DualOPL2 bit (newm): 1
2022-06-21 06:44:14.142 |  |- AdlibGold smart pointer: nullptr
2022-06-21 06:44:14.142 |  `- channel smart pointer: populated
2022-06-21 06:44:14.142 |       |- is enabled: no
2022-06-21 06:44:14.142 |       `- sample rate: 48000 Hz
```

#### `Z:\> oplmode opl3`

```
2022-06-21 06:44:18.381 | OPL: Diagnostics
2022-06-21 06:44:18.381 |  |- Mode: OPL3
2022-06-21 06:44:18.381 |  |- Port 388h: answering
2022-06-21 06:44:18.382 |  |- Port 220h is: answering
2022-06-21 06:44:18.382 |  |- Port 228h is: answering
2022-06-21 06:44:18.382 |  |- (prior) Port 220h is: answering
2022-06-21 06:44:18.382 |  |- (prior) Port 228h is: answering
2022-06-21 06:44:18.382 |  |- DualOPL2 bit (newm): 0
2022-06-21 06:44:18.382 |  |- AdlibGold smart pointer: nullptr
2022-06-21 06:44:18.382 |  `- channel smart pointer: populated
2022-06-21 06:44:18.383 |       |- is enabled: no
2022-06-21 06:44:18.383 |       `- sample rate: 48000 Hz
```

#### `Z:\> oplmode opl3gold`

```
2022-06-21 06:44:19.890 |  |- Mode: OPL3Gold
2022-06-21 06:44:19.890 |  |- Port 388h: answering
2022-06-21 06:44:19.891 |  |- Port 220h is: answering
2022-06-21 06:44:19.891 |  |- Port 228h is: answering
2022-06-21 06:44:19.891 |  |- (prior) Port 220h is: answering
2022-06-21 06:44:19.891 |  |- (prior) Port 228h is: answering
2022-06-21 06:44:19.891 |  |- DualOPL2 bit (newm): 0
2022-06-21 06:44:19.891 |  |- AdlibGold smart pointer: populated
2022-06-21 06:44:19.891 |  `- channel smart pointer: populated
2022-06-21 06:44:19.891 |       |- is enabled: no
2022-06-21 06:44:19.891 |       `- sample rate: 48000 Hz
```

#### `Z:\> sbbase 240`

```
2022-06-21 06:44:48.013 | OPL: Diagnostics
2022-06-21 06:44:48.013 |  |- Mode: OPL3Gold
2022-06-21 06:44:48.013 |  |- Port 388h: answering
2022-06-21 06:44:48.013 |  |- Port 240h is: answering
2022-06-21 06:44:48.013 |  |- Port 248h is: answering
2022-06-21 06:44:48.014 |  |- (prior) Port 220h is: not answering
2022-06-21 06:44:48.014 |  |- (prior) Port 228h is: not answering
2022-06-21 06:44:48.014 |  |- DualOPL2 bit (newm): 0
2022-06-21 06:44:48.014 |  |- AdlibGold smart pointer: populated
2022-06-21 06:44:48.015 |  `- channel smart pointer: populated
2022-06-21 06:44:48.015 |       |- is enabled: no
2022-06-21 06:44:48.015 |       `- sample rate: 48000 Hz
```

**Comment by kcgen** [comment]
I don't think this level of details needed (only added for my own confirmation locally), but a bit more differentiation detail would be appreciated :+1:

**Comment by johnnovak** [comment]
> I've got support for the OPL2LPT, Retrowave OPL, and the Linux kernel's OPL2 driver stack (for, e.g., real OPL2 soundcards), but it's all Linux-specific, and OPL2-only for now, so other OSes and OPL3 support are still a work-in-progress.

Yeah, so I think supporting MacOS, Linux and Windows is non-negotiable; I think introducing platform-specific features should be the last resort. That universal hardware OPL library of yours sounds like a nice idea, but then, the code in ScummVM is quite simple, so if you really wanted to do this today, you could just follow that same approach 😄 

------- 

Btw, I did listen to some comparison videos on YouTube out of sheer curiosity, and I couldn't tell the difference between the Retrowave and Nuked if my life depended on it. Which is unsurprising as it's a digital FM synth and Nuked is supposed to provide a bit-perfect emulation of the digital output. I guess the analog stage can make some minuscule difference, but then ruler flat D/As cost like 5 cents these days (literally). And there's no other analog component in it (e.g. filters with weird distortion characteristics like in the SID), so...

Of course, people claimed in the comments that the Retrowave is livelier/beefier/fuller/more organic/warmer/_[insert any vague audiophile term here]_... Let's just say I'm extremely sceptical of such claims, I doubt anyone could pick them apart in A/B tests.

My personal view: I consider Nuked to be the endgame of emulating OPL audio 😁🤘🏻

**Comment by johnnovak** [comment]
> I don't think this level of details needed (only added for my own confirmation locally), but a bit more differentiation detail would be appreciated 👍

Yeah this is existing behaviour, but adding a single log entry for the selected OPL mode would be helpful. I'll do that in my next PR, or feel free to add it yourself if you feel like it.

Nice diagnostic function, btw! (but we don't need that level of detail in the normal logs, I agree.)

**Comment by johnnovak** [comment]
Logging added in https://github.com/dosbox-staging/dosbox-staging/pull/1742

**Comment by kcgen** [comment]
Looking good so far! For now, the proof of concept integration is understandable enough that we know how complex it is. 

 > Linux only so far, but I'll try to expand on that for the final version.

Once the library is cross platform, I think then it will make sense for us collectively to put effort into the finishing touches of how it will be integrated.  You're questions are good - and they can wait here until then.

(not trying to be dismissive.. just prioritizing based on urgency and time-frames; sorting this out will take time and effort, and so we might as well move that activity to when the PR is a reality.  Thanks, @sulix )

**Comment by johnnovak** [comment]
> * The mixer channel won't affect a separate hardware OPL2, but it's needed as part of the emulation, so it becomes useless.

Now that we have a way of tagging mixer channels with "features", we could add a "Hardware" feature, and hide hardware channels in the mixer.

> * a single string "oplhw" option which takes the device path.

The less config options, the better. We can work this out later, but remember, "device path" sounds like a Linux only approach. This has to be a cross-platform feature (well, maybe some config options would only make sense on Linux, but the general structure has to be designed with cross-platform support in mind).

> * The code diff can be pretty small, albeit with some ugly "else" statements behind #ifdefs.

No, that definitely won't fly. You've effectively made this a compilation-time only option; it has to be selectable at runtime.

It seems you're the only person so far interested in adding support to the hardware OPL stuff, so you'll need to find a way to test your various hardware on Windows and macOS as well, which can take a while I guess. So let's get back to it when that has materialised, because that would definitely complicate the code a lot.

**Comment by sulix** [comment]
Thanks for taking a look — as you note, most of this discussion will be better suited to the final Pull Request, so I'll leave the detailed discussion until then. (As you note, there's a fair bit more polish needed, and only so many hours in a day, so it'll be a little while before it's ready.)

To quickly respond to the comments above:
- Tagging mixer channels sounds like an interesting approach: I'll look into it.
- The "device path" I'm thinking of would be (mostly) cross-platform. While the actual paths would often differ, the concept of some sort of "connection string" is valid across all OSes. In the current version, these encapsulate all of the required info, and would look something like "opl2lpt:LPT1" or "retrowave:COM1" or "ioport:388" on windows (to contrast "opl2lpt:/dev/parport0", "retrowave:/dev/ttyACM0", "ioport:388" on Linux).
- I agree the current proof-of-concept's hackiness with else and #ifdef should go, though I think it makes the POC diff a bit smaller and easier to read to illustrate the idea. My plan for the final version was to tidy up the implementation to not do that, but to still have it behind both compile and runtime options. That way, it'd be possible to compile without OPL hardware support if, e.g., the library were not available. If the compile-time option is enabled, the feature can be enabled/disabled at runtime in the dosbox-staging.conf. This mirrors the ALSA MIDI support, and makes liboplhw an optional dependency.
- OS-support-wise, Linux has kernel-level OPL2/3 APIs, so made sense to start with, but windows will definitely see full support. (It builds at the moment, but I haven't had a chance to set up a test environment yet.) macOS will be trickier for me, not having easy access to physical hardware to test against, but I'll do my best, and will definitely make sure it doesn't have any regressions, even if there's slightly less broad support for the more esoteric hardware setups.
- I did look at the ScummVM support, and decided to go with a new library for a couple of reasons. Firstly, I wanted to have a common way/API for supporting several of these devices across several projects (source ports, etc.). Secondly, I wanted to avoid adding a dependency on a new library for each new supported bit of hardware (particularly since the RetroWave library is under AGPL3, which is too restrictive to be used in several projects). And finally, ScummVM didn't have RetroWave support when I first implemented it in my own library. That being said, if you have a strong preference to use the various libraries ScummVM used, that's possible too — it'll just incur more dependencies and be a bit more invasive in the dosbox-staging code.

Thanks again: I'll send a PR in when things are more polished.

**Comment by johnnovak** [comment]
> * My plan for the final version was to tidy up the implementation to not do that, but to still have it behind both compile and runtime options.

> * This mirrors the ALSA MIDI support, and makes liboplhw an optional dependency.

Probably a good idea.

> * And finally, ScummVM didn't have RetroWave support when I first implemented it in my own library. That being said, if you have a strong preference to use the various libraries ScummVM used, that's possible too — it'll just incur more dependencies and be a bit more invasive in the dosbox-staging code.

No strong preference on my part, I was thinking about maintainability. E.g. because this is such a niche thing, I don't expect many people to be interested in it (you need to own the hardware in the first place, which I think reduces the potential audience to >0.1% of DOSBox users, and maybe I'm optimistic). That means that if you abandon your library at some point, then that's a bit of a problem as probably not too many people will be interested in maintaining it.

I guess by copying the ScummVM approach the risk is less because as long as hardware support is being maintained in that project, we could always just take their updates relatively easily.

In any case, up to you; as long as it doesn't complicate the normal emulated path I think you can do whatever 😄 

Out of curiosity, can you notice any tangible difference between your hardware OPL stuff and Nuked? Like under controlled ABX tests as I'm sure the _thought_ of using the hardware can make feel one a lot better (been there, I've "heard" all sorts of things during my "audio career", including tweaking bypassed plugins for minutes and hearing subtle differences... 🤣)

I'm just curious because in case of C64 SID music I can see some value in using hardware solutions as the SID had an analogue filter stage (however, the filter has been adequately modeled to a very high accuracy years ago; not that accuracy matters too much because almost every single C64 had a different analog filter due to quite low manufacturing tolerances...). But in case of the OPL, that's just a digital FM IC followed by a DAC, plus a very simple uncharacterful analog lowpass in some SB models or clones.

I've heard some recordings of the original Adlib on YouTube which distorts the audio like crazy; now I would see the value in owning that hardware, those distortions introduced by the analog circuitry can sound quite characterful indeed. But later OPL implementations found in SB boards and clones sound crystal clean to me, based on random recordings I found online.

**Comment by johnnovak** [comment]
The work for this has been completed. Added "documentation" tag & keeping the ticket open for documentation purposes.

**Comment by johnnovak** [comment]
@sulix Please open a separate ticket/PR for the hardware OPL support if you decided to add it to Staging.
